File: send_fcm_notification.js
/**
 * FCM Push Notification Sender - Node.js Example
 * ===============================================
 * 
 * Script untuk mengirim push notification FCM ke Flutter app.
 * Cocok untuk backend Node.js/Express implementation.
 * 
 * Installation:
 *   npm install axios
 * 
 * Usage:
 *   node send_fcm_notification.js
 */

const axios = require('axios');

// ==================== CONFIGURATION ====================
// Dapatkan dari Firebase Console → Project Settings → Cloud Messaging
const FCM_SERVER_KEY = 'YOUR_FCM_SERVER_KEY_HERE';

// FCM Token dari device (lihat di console Flutter app)
const DEVICE_TOKEN = 'DEVICE_FCM_TOKEN_HERE';

const FCM_URL = 'https://fcm.googleapis.com/fcm/send';
// =======================================================


/**
 * Kirim notifikasi promo ke device
 */
async function sendPromoNotification({
  token,
  title,
  body,
  promoId = null,
  promoType = 'promo'
}) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `key=${FCM_SERVER_KEY}`
  };

  const data = { type: promoType };
  if (promoId) {
    data.promo_id = promoId;
  }

  const payload = {
    to: token,
    notification: {
      title: title,
      body: body,
      sound: 'hidupjokowi',
      click_action: 'FLUTTER_NOTIFICATION_CLICK',
      priority: 'high'
    },
    data: data,
    priority: 'high'
  };

  try {
    const response = await axios.post(FCM_URL, payload, { headers });
    console.log('✅ Notifikasi berhasil dikirim!');
    console.log('Response:', JSON.stringify(response.data, null, 2));
    return response.data;
  } catch (error) {
    console.error('❌ Error mengirim notifikasi:', error.message);
    if (error.response) {
      console.error('Response:', error.response.data);
    }
    return null;
  }
}


/**
 * Kirim notifikasi update order
 */
async function sendOrderNotification(token, orderId, status) {
  const statusMessages = {
    'processing': 'Pesanan Anda sedang diproses',
    'cooking': 'Pesanan Anda sedang dimasak',
    'delivery': 'Pesanan Anda sedang diantar',
    'completed': 'Pesanan Anda telah selesai'
  };

  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `key=${FCM_SERVER_KEY}`
  };

  const payload = {
    to: token,
    notification: {
      title: 'Update Pesanan',
      body: statusMessages[status] || 'Status pesanan berubah',
      sound: 'hidupjokowi'
    },
    data: {
      type: 'order',
      order_id: orderId,
      status: status
    },
    priority: 'high'
  };

  try {
    const response = await axios.post(FCM_URL, payload, { headers });
    console.log('✅ Notifikasi order berhasil dikirim!');
    return response.data;
  } catch (error) {
    console.error('❌ Error:', error.message);
    return null;
  }
}


/**
 * Kirim notifikasi ke multiple devices
 */
async function sendBulkNotification(tokens, title, body, data = null) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `key=${FCM_SERVER_KEY}`
  };

  const payload = {
    registration_ids: tokens,
    notification: {
      title: title,
      body: body,
      sound: 'hidupjokowi'
    },
    data: data || { type: 'promo' },
    priority: 'high'
  };

  try {
    const response = await axios.post(FCM_URL, payload, { headers });
    const { success = 0, failure = 0 } = response.data;
    
    console.log('✅ Notifikasi bulk berhasil!');
    console.log(`   Success: ${success}`);
    console.log(`   Failed: ${failure}`);
    
    return response.data;
  } catch (error) {
    console.error('❌ Error:', error.message);
    return null;
  }
}


/**
 * Express.js Route Example
 */
function setupExpressRoutes(app) {
  // POST /api/send-promo-notification
  app.post('/api/send-promo-notification', async (req, res) => {
    const { token, title, body, promoId } = req.body;
    
    if (!token || !title || !body) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields'
      });
    }

    const result = await sendPromoNotification({
      token,
      title,
      body,
      promoId
    });

    if (result) {
      res.json({ success: true, data: result });
    } else {
      res.status(500).json({ success: false, message: 'Failed to send notification' });
    }
  });

  // POST /api/send-order-notification
  app.post('/api/send-order-notification', async (req, res) => {
    const { token, orderId, status } = req.body;
    
    if (!token || !orderId || !status) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields'
      });
    }

    const result = await sendOrderNotification(token, orderId, status);

    if (result) {
      res.json({ success: true, data: result });
    } else {
      res.status(500).json({ success: false, message: 'Failed to send notification' });
    }
  });
}


// ==================== CONTOH PENGGUNAAN ====================

async function main() {
  console.log('='.repeat(60));
  console.log('FCM Push Notification Sender');
  console.log('='.repeat(60));

  // Contoh 1: Kirim notifikasi promo umum
  console.log('\n1. Mengirim notifikasi promo umum...');
  await sendPromoNotification({
    token: DEVICE_TOKEN,
    title: '🎉 Promo Spesial Hari Ini!',
    body: 'Lihat semua promo menarik yang tersedia untuk Anda'
  });

  // Contoh 2: Kirim notifikasi promo spesifik
  console.log('\n2. Mengirim notifikasi promo spesifik...');
  await sendPromoNotification({
    token: DEVICE_TOKEN,
    title: '🔥 Diskon 50% Menu Spesial!',
    body: 'Gunakan kode SPESIAL50 untuk diskon hingga 50%',
    promoId: '1'
  });

  // Contoh 3: Kirim notifikasi order
  console.log('\n3. Mengirim notifikasi update order...');
  await sendOrderNotification(
    DEVICE_TOKEN,
    '12345',
    'delivery'
  );

  // Contoh 4: Kirim ke multiple devices
  console.log('\n4. Mengirim ke multiple devices...');
  // await sendBulkNotification(
  //   [DEVICE_TOKEN, 'token2', 'token3'],
  //   'Promo Weekend!',
  //   'Nikmati promo weekend spesial',
  //   { type: 'promo' }
  // );

  console.log('\n' + '='.repeat(60));
  console.log('Selesai!');
  console.log('='.repeat(60));
}

// Run if called directly
if (require.main === module) {
  main();
}

// Export untuk digunakan di aplikasi Express
module.exports = {
  sendPromoNotification,
  sendOrderNotification,
  sendBulkNotification,
  setupExpressRoutes
};

File: send_fcm_notification.py
"""
FCM Push Notification Sender - Python Example
==============================================

Script untuk mengirim push notification FCM ke Flutter app.
Cocok untuk backend/server-side implementation.

Requirements:
    pip install requests

Usage:
    python send_fcm_notification.py
"""

import requests
import json

# ==================== CONFIGURATION ====================
# Dapatkan dari Firebase Console → Project Settings → Cloud Messaging
FCM_SERVER_KEY = "YOUR_FCM_SERVER_KEY_HERE"

# FCM Token dari device (lihat di console Flutter app)
DEVICE_TOKEN = "DEVICE_FCM_TOKEN_HERE"

FCM_URL = "https://fcm.googleapis.com/fcm/send"
# =======================================================


def send_promo_notification(
    token: str,
    title: str,
    body: str,
    promo_id: str = None,
    promo_type: str = "promo"
):
    """
    Kirim notifikasi promo ke device
    
    Args:
        token: FCM token device
        title: Judul notifikasi
        body: Isi notifikasi
        promo_id: ID promo spesifik (optional, jika kosong akan ke list promo)
        promo_type: Tipe notifikasi (default: "promo")
    """
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"key={FCM_SERVER_KEY}"
    }
    
    # Data payload untuk navigasi
    data = {
        "type": promo_type
    }
    
    if promo_id:
        data["promo_id"] = promo_id
    
    payload = {
        "to": token,
        "notification": {
            "title": title,
            "body": body,
            "sound": "hidupjokowi",  # Custom sound (optional)
            "click_action": "FLUTTER_NOTIFICATION_CLICK",
            "priority": "high"
        },
        "data": data,
        "priority": "high"
    }
    
    try:
        response = requests.post(FCM_URL, headers=headers, data=json.dumps(payload))
        response.raise_for_status()
        
        result = response.json()
        print(f"✅ Notifikasi berhasil dikirim!")
        print(f"Response: {json.dumps(result, indent=2)}")
        return result
        
    except requests.exceptions.RequestException as e:
        print(f"❌ Error mengirim notifikasi: {e}")
        if hasattr(e.response, 'text'):
            print(f"Response: {e.response.text}")
        return None


def send_order_notification(token: str, order_id: str, status: str):
    """Kirim notifikasi update order"""
    
    status_messages = {
        "processing": "Pesanan Anda sedang diproses",
        "cooking": "Pesanan Anda sedang dimasak",
        "delivery": "Pesanan Anda sedang diantar",
        "completed": "Pesanan Anda telah selesai"
    }
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"key={FCM_SERVER_KEY}"
    }
    
    payload = {
        "to": token,
        "notification": {
            "title": "Update Pesanan",
            "body": status_messages.get(status, "Status pesanan berubah"),
            "sound": "hidupjokowi"
        },
        "data": {
            "type": "order",
            "order_id": order_id,
            "status": status
        },
        "priority": "high"
    }
    
    try:
        response = requests.post(FCM_URL, headers=headers, data=json.dumps(payload))
        response.raise_for_status()
        print(f"✅ Notifikasi order berhasil dikirim!")
        return response.json()
    except Exception as e:
        print(f"❌ Error: {e}")
        return None


def send_bulk_notification(tokens: list, title: str, body: str, data: dict = None):
    """
    Kirim notifikasi ke multiple devices
    
    Args:
        tokens: List of FCM tokens
        title: Judul notifikasi
        body: Isi notifikasi
        data: Custom data payload
    """
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"key={FCM_SERVER_KEY}"
    }
    
    payload = {
        "registration_ids": tokens,  # Multiple tokens
        "notification": {
            "title": title,
            "body": body,
            "sound": "hidupjokowi"
        },
        "data": data or {"type": "promo"},
        "priority": "high"
    }
    
    try:
        response = requests.post(FCM_URL, headers=headers, data=json.dumps(payload))
        response.raise_for_status()
        
        result = response.json()
        success_count = result.get('success', 0)
        failure_count = result.get('failure', 0)
        
        print(f"✅ Notifikasi bulk berhasil!")
        print(f"   Success: {success_count}")
        print(f"   Failed: {failure_count}")
        
        return result
        
    except Exception as e:
        print(f"❌ Error: {e}")
        return None


# ==================== CONTOH PENGGUNAAN ====================

if __name__ == "__main__":
    print("=" * 60)
    print("FCM Push Notification Sender")
    print("=" * 60)
    
    # Contoh 1: Kirim notifikasi promo umum (ke list promo)
    print("\n1. Mengirim notifikasi promo umum...")
    send_promo_notification(
        token=DEVICE_TOKEN,
        title="🎉 Promo Spesial Hari Ini!",
        body="Lihat semua promo menarik yang tersedia untuk Anda"
    )
    
    # Contoh 2: Kirim notifikasi promo spesifik (ke detail promo)
    print("\n2. Mengirim notifikasi promo spesifik...")
    send_promo_notification(
        token=DEVICE_TOKEN,
        title="🔥 Diskon 50% Menu Spesial!",
        body="Gunakan kode SPESIAL50 untuk diskon hingga 50%",
        promo_id="1"  # ID promo spesifik
    )
    
    # Contoh 3: Kirim notifikasi order
    print("\n3. Mengirim notifikasi update order...")
    send_order_notification(
        token=DEVICE_TOKEN,
        order_id="12345",
        status="delivery"
    )
    
    # Contoh 4: Kirim ke multiple devices
    print("\n4. Mengirim ke multiple devices...")
    # send_bulk_notification(
    #     tokens=[DEVICE_TOKEN, "token2", "token3"],
    #     title="Promo Weekend!",
    #     body="Nikmati promo weekend spesial",
    #     data={"type": "promo"}
    # )
    
    print("\n" + "=" * 60)
    print("Selesai!")
    print("=" * 60)
